src/string.cr

```diff
@@ -578,31 +578,58 @@ class String
     ToU64Info.new value, negative, invalid
   end
 
-  # Returns the result of interpreting leading characters in this string as a floating point number (`Float64`).
-  # Extraneous characters past the end of a valid number are ignored. If there is not a valid number at the start of str,
-  # 0.0 is returned. This method never raises an exception.
+  # Returns the result of interpreting characters in this string as a floating point number (`Float64`).
+  # This method raises an exception if the string is not a valid float representation.
   #
   # ```
   # "123.45e1".to_f      # => 1234.5
   # "45.67 degrees".to_f # => 45.67
-  # "thx1138".to_f       # => 0.0
+  # "thx1138".to_f       # => ArgumentError
   # ```
   def to_f
     to_f64
   end
 
-  # Returns the result of interpreting leading characters in this string as a floating point number (`Float32`).
-  # Extraneous characters past the end of a valid number are ignored. If there is not a valid number at the start of str,
-  # 0.0 is returned. This method never raises an exception.
+  # Returns the result of interpreting characters in this string as a floating point number (`Float64`).
+  # This method returns `nil` if the string is not a valid float representation.
+  #
+  # ```
+  # "123.45e1".to_f?      # => 1234.5
+  # "45.67 degrees".to_f? # => 45.67
+  # "thx1138".to_f?       # => nil
+  # ```
+  def to_f?
+    to_f64?
+  end
+
+  # Returns the result of interpreting characters in this string as a floating point number (`Float32`).
+  # This method raises an exception if the string is not a valid float representation.
   #
   # See `#to_f`.
   def to_f32
-    LibC.strtof self, nil
+    to_f32? || raise ArgumentError.new("Invalid Float32: #{self}")
+  end
+
+  # Returns the result of interpreting characters in this string as a floating point number (`Float32`).
+  # This method returns `nil` if the string is not a valid float representation.
+  #
+  # See `#to_f`.
+  def to_f32?
+    return nil unless '0' <= self[0] <= '9' || self[0] == '-' || self[0] == '+'
+    v = LibC.strtof self, out endptr
+    endptr == (to_unsafe + bytesize) ? v : nil
   end
 
   # Same as `#to_f`.
   def to_f64
-    LibC.atof self
+    to_f64? || raise ArgumentError.new("Invalid Float64: #{self}")
+  end
+
+  # Same as `#to_f?`.
+  def to_f64?
+    return nil unless '0' <= self[0] <= '9' || self[0] == '-' || self[0] == '+'
```

Comment:
Why is this check needed? I imagine because leading whitespace is allowed by `strtod`.

Eventually we should have some of the options as `to_i`: 
* **whitespace** (defaults to `true`): if true, leading and trailing whitespaces are allowed
* **strict** (defaults to `true`): if true, extraneous characters past the end of the number are disallowed